core/common/src/main/java/alluxio/concurrent/CountLatch.java

```diff
@@ -0,0 +1,155 @@
+ * (the "License"). You may not use this work except in compliance with the License, which is
+ * available at www.apache.org/licenses/LICENSE-2.0
+ *
+ * This software is distributed on an "AS IS" basis, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied, as more fully set forth in the License.
+ *
+ * See the NOTICE file distributed with this work for information regarding copyright ownership.
+ */
+
+package alluxio.concurrent;
+
+import com.google.common.annotations.VisibleForTesting;
+
+import java.util.concurrent.locks.AbstractQueuedSynchronizer;
+
+/**
+ * Similar to {@link java.util.concurrent.CountDownLatch} except that the count can either be
+ * increased or decreased.
+ *
+ * A thread can wait until the count reaches 0.
+ * Other threads can increase or decrease the count concurrently.
+ * When count reaches 0, if there is a thread waiting for state 0, the thread will be waken up,
+ * and other threads trying to increase the count are blocked. If there is no thread waiting for
+ * state 0, then other threads can keep increasing the count.
+ *
+ * Methods can be run concurrently from different threads, the paired methods({@link #inc()} and
+ * {@link #dec()}, {@link #awaitAndBlockInc()} and {@link #unblockInc()}) do not need to be
+ * called in the same thread.
+ *
+ * An example usage is to synchronize FileSystemContext reinitialization with the ongoing RPCs.
+ * See alluxio.client.fs.FileSystemContextReinitializer for more details.
+ */
+public class CountLatch {
+  /** Represents the ignored arguments for AQS methods. */
+  private static final int IGNORED_ARG = -1;
+  /** Represents the state where {@link #inc()} is blocked. */
+  private static final int BLOCK_INC_STATE = -1;
+
+  /**
+   * Meaning of the AQS state:
+   * 1. when equals BLOCK_INC_STATE, means awaitAndBlockInc has returned, and all further calls to
+   * inc are blocked until unblockInc is called.
+   * 2. when >= 0, means the counter value increased by inc and decreased by dec.
+   */
+  private static final class Sync extends AbstractQueuedSynchronizer {
+    private static final long serialVersionUID = 8553010505281724322L;
+
+    @Override
+    protected boolean tryAcquire(int arg) {
+      while (true) {
+        if (getState() != 0) {
+          return false;
+        }
+        if (compareAndSetState(0, BLOCK_INC_STATE)) {
+          return true;
+        }
+      }
+    }
+
+    @Override
+    protected boolean tryReleaseShared(int arg) {
+      if (getState() != BLOCK_INC_STATE || !compareAndSetState(BLOCK_INC_STATE, 0)) {
+        throw new Error("state must be " + BLOCK_INC_STATE);
+      }
+      return true;
+    }
+
+    @Override
+    protected int tryAcquireShared(int arg) {
+      while (true) {
+        int current = getState();
+        if (current == BLOCK_INC_STATE) {
+          return -1;
+        }
+        if (compareAndSetState(current, current + 1)) {
```

Comment:
What if the block state was cleared after calling getState()? Wouldn't this thread sleep until next time since shared waiters has been woken up alredy?

Reply:
if the block state is cleared after `getState`, then compareAndSetState will fail, so the while loop will continue, in the next loop, getState will no longer be block state, so  the thread won't go to sleep.